src/core/unif.ml

```diff
@@ -207,9 +265,9 @@ and solve_aux : ctxt -> term -> term -> problem -> constr list =
 
   let decompose () =
     (* Propagate context *)
-    let add_arg_pb l a b = (ctx,a,b)::l in
+    let add_arg_pb a b l = (ctx,a,b)::l in
     let to_solve =
-      try List.fold_left2 add_arg_pb p.to_solve ts1 ts2
+      try List.fold_right2 add_arg_pb ts1 ts2 p.to_solve
```

Comment:
Why do you change fold_left to fold_right (which is not tail-recursive)?

Reply:
here the rationale is to decompose f a b = f a' b' to a = a'  and b = b' instead of b = b' and a = a'

keeping the order, the algorithm study the equations related to implicit arguments first

Reply:
Thanks to add a comment here. Indeed, with dependent types, it is a good idea to keep the order.

Reply:
fixed